runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/interceptors/InterceptorExecutor.kt

```diff
@@ -0,0 +1,326 @@
+    override var protocolRequest: HttpRequest,
+    private val executionContext: ExecutionContext,
+) : ProtocolRequestInterceptorContext<I, HttpRequest>, Attributes by executionContext
+
+private data class HttpProtocolResponseInterceptorContext<I>(
+    override val request: I,
+    override val protocolRequest: HttpRequest,
+    override var protocolResponse: HttpResponse,
+    private val executionContext: ExecutionContext,
+) : ProtocolResponseInterceptorContext<I, HttpRequest, HttpResponse>, Attributes by executionContext
+
+private data class HttpAttemptInterceptorContext<I, O>(
+    override val request: I,
+    override var response: Result<O>,
+    override val protocolRequest: HttpRequest,
+    override val protocolResponse: HttpResponse?,
+    private val executionContext: ExecutionContext,
+) : ResponseInterceptorContext<I, O, HttpRequest, HttpResponse?>, Attributes by executionContext
+
+private data class HttpInputOutputInterceptorContext<I, O>(
+    override val request: I,
+    override var response: Result<O>,
+    private val call: HttpCall,
+    private val executionContext: ExecutionContext,
+) : ResponseInterceptorContext<I, O, HttpRequest, HttpResponse>, Attributes by executionContext {
+
+    override val protocolRequest: HttpRequest = call.request
+    override val protocolResponse: HttpResponse = call.response
+}
+
+private data class HttpFinalInterceptorContext<I, O>(
+    override val request: I,
+    override var response: Result<O>,
+    override val protocolRequest: HttpRequest?,
+    override val protocolResponse: HttpResponse?,
+    private val executionContext: ExecutionContext,
+) : ResponseInterceptorContext<I, O, HttpRequest?, HttpResponse?>, Attributes by executionContext
+
+// TODO - investigate propagating Any as upper bounds for SdkHttpOperation <I,O> generics
+/**
+ * Fan-out facade over raw interceptors that adapts the internal view of an operation execution to the public facing
+ * view interceptor hooks see.
+ *
+ * @param execContext the operation [ExecutionContext]
+ * @param interceptors the list of interceptors to execute
+ * @param typeInfo the expected input/output type info used to keep hook implementations honest
+ */
+internal class InterceptorExecutor<I, O>(
+    private val execContext: ExecutionContext,
+    private val interceptors: List<HttpInterceptor>,
+    private val typeInfo: OperationTypeInfo,
+) {
+
+    // This changes as execution progresses, it represents the most up-to-date version of the operation input.
+    // If we begin executing an operation at all it is guaranteed to exist because `readBeforeExecution` is the first
+    // thing invoked when executing a request
+    private var _lastInput: I? = null
+
+    // Track most up to date http request and response. The final two hooks do not have easy access to this data
+    // so we store it as execution progresses
+    private var _lastHttpRequest: HttpRequest? = null
+    private var _lastHttpResponse: HttpResponse? = null
```

Comment:
`InterceptorExecutor` is scoped to a single operation. All the hooks should fire from one thread (may not be the same thread for the entirety of the operation because coroutines can switch after suspension points but only one thread should ever be interacting with it).

Is there a scenario where you see this as not holding true?

Reply:
No, that makes sense.